File: ts/isValidParentheses.ts
class Stack {
  private state: string[] = [];

  push(char: string) {
    if (char === ')') return this.handleA(char);
    if (char === '}') return this.handleB(char);
    if (char === ']') return this.handleC(char);

    this.handleOpeners(char);
  }

  private handleOpeners(char: string) {
    this.state.push(char);
    this.logState();
  }

  private handleA(char: string) {
    if (this.last === '(') {
      this.state.pop();
    } else {
      this.state.push(char);
    }

    this.logState();
  }

  private handleB(char: string) {
    if (this.last === '{') {
      this.state.pop();
    } else {
      this.state.push(char);
    }
    this.logState();
  }

  private handleC(char: string) {
    if (this.last === '[') {
      this.state.pop();
    } else {
      this.state.push(char);
    }
    this.logState();
  }

  private logState() {
    console.log(this.state);
  }

  public get length() {
    return this.state.length;
  }

  private get last() {
    return this.state.at(this.state.length - 1);
  }
}

const brackets = {
  ')': '(',
  '}': '{',
  ']': '[',
} as const;
const openers = ['(', '{', '['] as const;
const stack: string[] = [];

function isValidParentheses(s: string): boolean {
  for (let i = 0; i < s.length; i++) {
    const bracket = s[i] as keyof typeof brackets &
      (typeof brackets)[keyof typeof brackets];

    if (openers.includes(bracket)) {
      stack.push(bracket);
    } else if (stack.pop() !== brackets[bracket]) {
      return false;
    }
  }

  return stack.length === 0;
}

export default () => {
  const s = '{[]}';
  // const s = '(])';
  // const s = '(]';
  // const s = '()[]{}';
  // const s = '()';
  const result = isValidParentheses(s);
  console.log(result);
};

File: rs/src/is_valid_parentheses.rs
use std::collections::HashMap;

fn is_valid_parentheses(s: String) -> bool {
    let brackets = HashMap::from([(')', '('), ('}', '{'), (']', '[')]);
    let mut stack: Vec<char> = vec![];
    let openers = ['(', '{', '['];

    for bracket in s.chars().into_iter() {
        println!("stack: {:?} | bracket: {bracket}", stack);
        if openers.contains(&bracket) {
            stack.push(bracket);
        } else {
            match stack.pop() {
                Some(last) => {
                    let &matching = brackets.get(&bracket).unwrap();

                    println!("last: {last} | matching: {matching}");

                    if last != matching {
                        return false;
                    }
                }
                None => {
                    return false;
                }
            };
        }
    }

    return stack.len() == 0;
}

pub fn main() {
    let s: String = "()[]{}".to_string();
    // let s: String = "()".to_string();
    let result = is_valid_parentheses(s);

    println!("{result}");
}
